modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/CacheQueryPageRequester.java

```diff
@@ -0,0 +1,236 @@
+
+            sendRequest(fut, req, n);
+
+        } catch (IgniteCheckedException e) {
+            fut.onDone(e);
+        }
+    }
+
+    /**
+     * Send cancel query request, so no new pages will be sent.
+     *
+     * @param reqId Query request ID.
+     * @param nodes Collection of nodes to send the cancel request.
+     * @param fieldsQry Whether query is a fields query.
+     *
+     */
+    public void cancelQuery(long reqId, Collection<UUID> nodes, boolean fieldsQry) {
+        final GridCacheQueryManager qryMgr = cctx.queries();
+
+        assert qryMgr != null;
+
+        try {
+            final GridCacheQueryRequest req = new GridCacheQueryRequest(cctx.cacheId(),
+                reqId,
+                fieldsQry,
+                cctx.startTopologyVersion(),
+                cctx.deploymentEnabled());
+
+            for (UUID node : nodes) {
+                try {
+                    if (cctx.localNodeId().equals(node)) {
+                        sendLoc.accept(req);
+
+                        continue;
+                    }
+
+                    cctx.io().send(node, req, GridIoPolicy.QUERY_POOL);
+                }
+                catch (IgniteCheckedException e) {
+                    if (cctx.io().checkNodeLeft(node, e, false)) {
+                        if (log.isDebugEnabled())
+                            log.debug("Failed to send cancel request, node failed: " + node);
+                    }
+                    else
+                        U.error(log, "Failed to send cancel request [node=" + node + ']', e);
+                }
+            }
+        }
+        catch (IgniteCheckedException e) {
+            U.error(log, "Failed to send cancel request (will cancel query in any case).", e);
+        }
+    }
+
+    /**
+     * Sends query request.
+     *
+     * @param req Request.
+     * @param nodes Nodes.
+     * @throws IgniteCheckedException In case of error.
+     */
+    private void sendRequest(
+        final GridCacheQueryFutureAdapter fut,
+        final GridCacheQueryRequest req,
+        Collection<ClusterNode> nodes
+    ) throws IgniteCheckedException {
+        assert req != null;
+        assert nodes != null;
+
+        final UUID locNodeId = cctx.localNodeId();
+
+        ClusterNode locNode = null;
+
+        Collection<ClusterNode> rmtNodes = null;
+
+        for (ClusterNode n : nodes) {
+            if (n.id().equals(locNodeId))
+                locNode = n;
+            else {
+                if (rmtNodes == null)
+                    rmtNodes = new ArrayList<>(nodes.size());
+
+                rmtNodes.add(n);
+            }
+        }
+
+        // Request should be sent to remote nodes before the query is processed on the local node.
+        // For example, a remote reducer has a state, we should not serialize and then send
+        // the reducer changed by the local node.
+        if (!F.isEmpty(rmtNodes)) {
+            for (ClusterNode node : rmtNodes) {
```

Comment:
```suggestion            
```

Reply:
I applied this suggestions, also I unified a code of sending cancel a new pages requests.